flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java

```diff
@@ -526,6 +526,16 @@ public void collect(StreamRecord<T> record) {
 				StreamRecord<T> copy = castRecord.copy(serializer.copy(castRecord.getValue()));
 				operator.setKeyContextElement1(copy);
 				operator.processElement(copy);
+			} catch (ClassCastException e) {
+				// Enrich error message
+				ClassCastException replace = new ClassCastException(
+						String.format("%s. Failed to push OutputTag with id '%s' to operator. " +
+								"This can occur when multiple OutputTags with different types " +
+								"but identical names are being used.",
+								e.getMessage(), outputTag.getId()));
+
+				throw new ExceptionInChainedOperatorException(replace);
```

Comment:
Swallows the original exception which is never a good thing for someone trying to diagnose a problem like this.